src/ray/gcs/redis_context.cc

```diff
@@ -45,74 +46,95 @@ namespace ray {
+                                                   << redisReply->str;
+  this->redis_reply_ = redisReply;
+}
+
+bool CallbackReply::IsNil() const { return REDIS_REPLY_NIL == redis_reply_->type; }
+
+uint64_t CallbackReply::ReadAsInteger() const {
+  RAY_CHECK(REDIS_REPLY_INTEGER == redis_reply_->type) << "Unexpected type:"
+                                                       << redis_reply_->type;
+  return static_cast<uint64_t>(redis_reply_->integer);
+}
+
+std::string CallbackReply::ReadAsString() const {
+  if (REDIS_REPLY_NIL == redis_reply_->type || REDIS_REPLY_STATUS == redis_reply_->type) {
+    return "";
   }
-  int64_t callback_index = reinterpret_cast<int64_t>(privdata);
-  redisReply *reply = reinterpret_cast<redisReply *>(r);
+
+  RAY_CHECK(REDIS_REPLY_STRING == redis_reply_->type) << "Unexpected type:"
+                                                      << redis_reply_->type;
+  return std::string(redis_reply_->str, redis_reply_->len);
+}
+
+Status CallbackReply::ReadAsStatus() const {
+  RAY_CHECK(REDIS_REPLY_STATUS == redis_reply_->type) << "Unexpected type:"
+                                                      << redis_reply_->type;
+  const std::string status_str(redis_reply_->str, redis_reply_->len);
+  if ("OK" == status_str) {
+    return Status::OK();
+  }
+
+  return Status::RedisError(status_str);
+}
+
+std::string CallbackReply::ReadAsPubsubData() const {
+  RAY_CHECK(REDIS_REPLY_ARRAY == redis_reply_->type) << "Unexpected type:"
+                                                     << redis_reply_->type;
+
   std::string data = "";
-  // Parse the response.
-  switch (reply->type) {
-  case (REDIS_REPLY_NIL): {
-    // Do not add any data for a nil response.
-  } break;
-  case (REDIS_REPLY_STRING): {
-    data = std::string(reply->str, reply->len);
-  } break;
-  case (REDIS_REPLY_STATUS): {
-  } break;
-  case (REDIS_REPLY_ERROR): {
-    RAY_LOG(FATAL) << "Redis error: " << reply->str;
-  } break;
-  case (REDIS_REPLY_INTEGER): {
-    data = std::to_string(reply->integer);
-    break;
+  // Parse the published message.
+  redisReply *message_type = redis_reply_->element[0];
+  if (strcmp(message_type->str, "subscribe") == 0) {
+    // If the message is for the initial subscription call, return the empty
+    // string as a response to signify that subscription was successful.
+  } else if (strcmp(message_type->str, "message") == 0) {
+    // If the message is from a PUBLISH, make sure the data is nonempty.
+    redisReply *message = redis_reply_->element[redis_reply_->elements - 1];
+    // data is a notification message.
+    data = std::string(message->str, message->len);
+    RAY_CHECK(!data.empty()) << "Empty message received on subscribe channel.";
+  } else {
+    RAY_LOG(FATAL) << "This is not a pubsub reply: data=" << message_type->str;
+  }
+
+  return data;
+}
+
+std::vector<std::string> CallbackReply::ReadAsStringArray() const {
```

Comment:
who's supposed to use this `ReadAsStringArray`?
Note, this function returns a copy. It could be an issue if the vector is large.

Reply:
> who's supposed to use this ReadAsStringArray?

It not be used in this PR, but I believe it'll be used when fetching a `set` from `GCS` soon.

> Note, this function returns a copy. It could be an issue if the vector is large.

Nice pointing! I'd like to make it as an output parameter.